### PR TITLE
Make DataValueDeserializer accept callbacks

### PR DESCRIPTION
**Pull request by thiemowmde**
Not an alternative but an addition to https://github.com/wmde/WikibaseDataModel/pull/669.

[Bug: T132592](https://phabricator.wikimedia.org/T132592)

### CODE REVIEW

**Comment by JeroenDeDauw** [comment]
I'd have added a second constructor parameter, though this also works. +1 to the approach

**Comment by thiemowmde** [comment]
I finished this patch, rewrote the most critical bits of documentation and removed the WIP. Please review carefully and merge. This should be a simple addition, nothing should break.

File: src/Deserializers/DataValueDeserializer.php [hunk 1]
```diff
@@ -15,45 +16,56 @@
...
 	const TYPE_KEY = 'type';
 	const VALUE_KEY = 'value';
 
 	/**
-	 * @var string[] Associative array of data type id => DataValue class name.
+	 * @var array Associative array mapping data type IDs to either callables returning new
+	 *  DataValue objects, or full qualified DataValue class names.
 	 */
-	private $classes;
+	private $builders;
```
Comment (brightbyte) [inline src/Deserializers/DataValueDeserializer.php]: "builder" is not a good term for this. $callbacks? $handlers?

File: src/Deserializers/DataValueDeserializer.php [hunk 2]
```diff
@@ -15,45 +16,56 @@
...
  */
 class DataValueDeserializer implements DispatchableDeserializer {
 
 	const TYPE_KEY = 'type';
 	const VALUE_KEY = 'value';
 
 	/**
-	 * @var string[] Associative array of data type id => DataValue class name.
+	 * @var array Associative array mapping data type IDs to either callables returning new
+	 *  DataValue objects, or full qualified DataValue class names.
```
Comment (brightbyte) [inline src/Deserializers/DataValueDeserializer.php]: What do the callbacks get as a parameter? If it's just a class name, what are the expectations? I assume such classes must have a static newFromArray method?

File: src/Deserializers/DataValueDeserializer.php (hunk 1 above)
Comment (JeroenDeDauw) [inline src/Deserializers/DataValueDeserializer.php]: I think $handlers is a lot worse than $builders, and find $callbacks less intent revealing

Really does not matter much IMO tho

**Pull request merged**